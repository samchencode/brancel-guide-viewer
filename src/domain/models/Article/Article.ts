import type { ArticleId } from '@/domain/models/Article/ArticleId';
import type { RichText } from '@/domain/models/RichText/RichText';

class Article {
  public readonly sectionIds: Set<string>;

  constructor(
    public readonly id: ArticleId,
    public readonly title: string,
    public readonly body: RichText,
    sectionIds: string[] = []
  ) {
    this.sectionIds = new Set(sectionIds).add(id.toString());
  }

  hasSection(id: string) {
    return this.sectionIds.has(id);
  }
}

export { Article };
