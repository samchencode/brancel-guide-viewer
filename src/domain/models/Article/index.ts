import type { ARTICLE_TYPES } from '@/domain/models/Article/constants';

type ValueOf<T> = T[keyof T];

export type ArticleType = ValueOf<typeof ARTICLE_TYPES>;

export { About } from '@/domain/models/Article/About';
export { Index } from '@/domain/models/Article/Index_';
export { UsageInstructions } from '@/domain/models/Article/UsageInstructions';
export { Article } from '@/domain/models/Article/Article';
export { ArticleId } from '@/domain/models/Article/ArticleId';
export type { ArticleRepository } from '@/domain/models/Article/ArticleRepository';
export type { ArticleCache } from '@/domain/models/Article/ArticleCache';
export type { ArticleRenderer } from '@/domain/models/Article/ArticleRenderer';
export { ARTICLE_TYPES } from '@/domain/models/Article/constants';
export { SearchableArticle } from '@/domain/models/Article/SearchableArticle';
export {
  ArticleSearchResult,
  ArticleMatchData,
  StringSegment,
} from '@/domain/models/Article/ArticleSearchResult';
export { ArticleSearch } from '@/domain/models/Article/ArticleSearch';
