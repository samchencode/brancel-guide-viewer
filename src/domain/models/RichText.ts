class RichText {
  constructor(
    private readonly sanitizeHtml: (html: string) => string,
    public readonly html: string
  ) {}

  getText() {
    return this.sanitizeHtml(this.html);
  }

  toString() {
    return this.html;
  }
}

export { RichText };
