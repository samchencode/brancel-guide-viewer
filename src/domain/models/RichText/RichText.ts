class RichText {
  constructor(public readonly html: string) {}

  toString() {
    return this.html;
  }
}

export { RichText };
