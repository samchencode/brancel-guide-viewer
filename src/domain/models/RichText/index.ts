export { RichText } from '@/domain/models/RichText/RichText';
export * from '@/domain/models/RichText/htmlManipulationUtils';
