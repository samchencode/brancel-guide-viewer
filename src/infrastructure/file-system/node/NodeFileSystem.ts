import fs from 'fs';
import path from 'path';
import type {
  EncodingType,
  FileSystem,
} from '@/infrastructure/file-system/FileSystem';
import { ENCODING_TYPES } from '@/infrastructure/file-system/constants';

const NODE_ENCODING_TYPES = {
  [ENCODING_TYPES.UTF_8]: 'utf-8',
  [ENCODING_TYPES.BASE_64]: 'base64',
} as const;

type ValueOf<T> = T[keyof T];

type NodeEncodingTypes = ValueOf<typeof NODE_ENCODING_TYPES>;

const readUtf8File = (p: string, encoding: NodeEncodingTypes) =>
  new Promise<string>((resolve, reject) => {
    fs.readFile(p, encoding, (err, data) =>
      err ? reject(err) : resolve(data)
    );
  });

class NodeFileSystem implements FileSystem<'jest'> {
  async getAssetAsString(
    aliasedFilePath: string,
    encodingType: EncodingType = 'utf8'
  ): Promise<string> {
    const filePath = aliasedFilePath.replace(
      /^@/,
      path.resolve(__dirname, '../../..')
    );
    return readUtf8File(filePath, NODE_ENCODING_TYPES[encodingType]);
  }

  cacheFile(): Promise<string> {
    throw new Error('Method not implemented.');
  }

  readFileAsString(): Promise<string> {
    throw new Error('Method not implemented.');
  }

  checkFileExists(): Promise<boolean> {
    throw new Error('Method not implemented.');
  }

  deleteFile(): Promise<void> {
    throw new Error('Method not implemented.');
  }
}

export { NodeFileSystem };
