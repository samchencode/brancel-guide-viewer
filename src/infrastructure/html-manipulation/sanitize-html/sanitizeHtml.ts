import type { SanitizeHtml } from '@/domain/models/RichText/htmlManipulationUtils';
import { sanitizeHtml as sanitize } from '@/vendor/sanitizeHtml';

export const sanitizeHtml: SanitizeHtml = sanitize;
