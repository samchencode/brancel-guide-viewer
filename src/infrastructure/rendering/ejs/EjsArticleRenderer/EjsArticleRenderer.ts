import { ejs } from '@/vendor/ejs';
import type { FileSystem } from '@/infrastructure/file-system/FileSystem';
import type { Article, ArticleRenderer } from '@/domain/models/Article';

class EjsArticleRenderer implements ArticleRenderer {
  prepareTemplate: Promise<ejs.TemplateFunction>;

  constructor(fileSystem: FileSystem) {
    this.prepareTemplate = this.init(fileSystem);
  }

  private async init(fileSystem: FileSystem) {
    const templateText = await fileSystem.getAssetAsString(
      require('@/infrastructure/rendering/ejs/EjsArticleRenderer/template.ejs')
    );
    return ejs.compile(templateText);
  }

  async render(a: Article) {
    const template = await this.prepareTemplate;
    return template({ title: a.title, body: a.body.html });
  }

  static $inject = ['fileSystem'];
}

export { EjsArticleRenderer };
