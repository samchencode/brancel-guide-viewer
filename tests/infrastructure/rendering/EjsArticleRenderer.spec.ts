import { Article, ArticleId } from '@/domain/models/Article';
import { RichText } from '@/domain/models/RichText';
import { NodeFileSystem } from '@/infrastructure/file-system/node/NodeFileSystem';
import { EjsArticleRenderer } from '@/infrastructure/rendering/ejs/EjsArticleRenderer/EjsArticleRenderer';
import { sanitizeHtml } from '@/vendor/sanitizeHtml';

describe('EjsArticleRenderer', () => {
  let fs: NodeFileSystem;
  beforeAll(() => {
    fs = new NodeFileSystem();
    jest.resetModules();
    jest.doMock(
      '@/infrastructure/rendering/ejs/EjsArticleRenderer/template.ejs',
      () => '@/infrastructure/rendering/ejs/EjsArticleRenderer/template.ejs'
    );
  });

  describe('Instantiation', () => {
    it('should create new renderer given fs', () => {
      const create = () => new EjsArticleRenderer(fs);
      expect(create).not.toThrowError();
    });
  });

  describe('Behavior', () => {
    it('should render html in the body', async () => {
      const id = new ArticleId('example');
      const title = 'Example Title';
      const body = new RichText(sanitizeHtml, '<h1>Hello World</h1>');
      const article = new Article(id, title, body);

      const renderer = new EjsArticleRenderer(fs);
      const html = await renderer.render(article);

      expect(html).not.toBe('');
      expect(html).toContain('<h1>Hello World</h1>');
      expect(html).toMatchSnapshot();
    });

    it('should not render html in the title', async () => {
      const id = new ArticleId('example');
      const title = '<h1>Example Title</h1>';
      const body = new RichText(sanitizeHtml, '<h1>Hello World</h1>');
      const article = new Article(id, title, body);

      const renderer = new EjsArticleRenderer(fs);
      const html = await renderer.render(article);

      expect(html).not.toContain('<h1>Example Title</h1>');
    });
  });
});
