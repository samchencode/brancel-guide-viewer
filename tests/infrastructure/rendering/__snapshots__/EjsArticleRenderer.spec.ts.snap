// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EjsArticleRenderer Behavior should render html in the body 1`] = `
"<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Example Title</title>
  <style>
    /* Global Styles */
    :root {
      font-family: 'Roboto', sans-serif;
    }

    * {
      margin: 0;
      padding: 0;
      box-sizing: border-box;
    }

    main {
      padding: 1em 1em;
      white-space: pre-wrap;
    }

    /* Text Styles */
    h2,
    h3 {
      white-space: pre-wrap;
    }

    p {
      white-space: pre-wrap;
    }

    /* Link Styles */
    a[href],
    a:visited {
      border-radius: .2em;
      padding: 0 .1em;
    }

    /* Table Styles */
    table {
      width: 100%;
    }

    th,
    td {
      padding: .3em .5em;
    }

    .table__wrapper {
      width: 100%;
      overflow-x: auto;
    }

    .table__wrapper::before {
      content: "(If this table doesn't fit on your screen, scroll sideways!)";
    }

    tr:nth-child(2n + 1):not(:only-child) td {
      background: lightgray;
    }

    /* Image Styles */
    img {
      width: 100%;
    }

    .img__wrapper {
      width: 100%;
    }

    .img__wrapper::after {
      content: "(If this image is too small, pinch to zoom!)";
    }

    /* List styles */
    ul,
    ol {
      padding-left: .5em;
      margin-left: 1em;
      white-space: normal;
    }

    li {
      white-space: pre-wrap;
      padding: 0.5em 0;
    }

    mark {
      margin: 0;
      padding: 0;
    }
  </style>
</head>

<body>
  <main id="body"><h1>Hello World</h1></main>
  <script>
    function notify(event) {
      var message = JSON.stringify(event);
      window.ReactNativeWebView.postMessage(message)
    }

    function notifyLinkPressed(href) {
      notify({
        type: 'linkpressed',
        data: {
          href: href,
        }
      })
    }

    function notifyError(name, msg) {
      notify({
        type: 'error',
        data: {
          name: name,
          message: msg,
        }
      })
    }

    function notifyIndexPressed() {
      notify({
        type: 'indexpressed',
        data: {}
      })
    }

    function notifyLog(msg) {
      notify({
        type: 'log',
        data: {
          message: msg
        }
      })
    }

    function notifyTableOfContentsPressed() {
      notify({
        type: 'tableofcontentspressed',
        data: {}
      })
    }

    ; // capture link presses
    (function() {
      var INDEX_ID = '#INDEX'
      var TABLE_OF_CONTENTS_ID = '#TABLE_OF_CONTENTS'

      var anchors = document.querySelectorAll('a[href]');
      for (var i = 0; i < anchors.length; i++) {
        anchors[i].addEventListener('click', function(e) {
          e.preventDefault();
          var url = new URL(this.href);
          if (url.hash === INDEX_ID) notifyIndexPressed();
          else if (url.hash === TABLE_OF_CONTENTS_ID) notifyTableOfContentsPressed();
          else notifyLinkPressed(this.href);
        })
      }
    })()

    ; // capture errors
    (function() {
      window.addEventListener('error', (e) => {
        notifyError(e.error.name, e.message);
      })
    })()

    ; // wrap tables
    (function() {
      var tables = document.getElementsByTagName('table');
      for (var i = 0; i < tables.length; i++) {
        var wrapper = document.createElement('div');
        wrapper.className = "table__wrapper";
        wrapper.appendChild(tables[i].cloneNode(true));
        tables[i].replaceWith(wrapper);
      }
    })()

    ; // wrap images
    (function() {
      var images = document.getElementsByTagName('img');
      for (var i = 0; i < images.length; i++) {
        var wrapper = document.createElement('div');
        wrapper.className = "img__wrapper";
        wrapper.appendChild(images[i].cloneNode());
        images[i].replaceWith(wrapper);
      }
    })()

    ; // get all text nodes to highlight if they match a search term
    TextMarker = (function() {
      function htmlEncode(input) {
        const textArea = document.createElement("textarea");
        textArea.innerText = input;
        return textArea.innerHTML.split("<br>").join("\\n");
      }

      function htmlDecode(input) {
        var textarea = document.createElement("textarea");
        textarea.innerHTML = input;
        return textarea.value;
      }

      /** 
       * @typedef Range
       * @type {[number, number]}
       */

      /** 
       * @typedef OverlapRange
       * @type {object}
       * @property {Range} range
       * @property {Range} overlap
       */

      /** 
       * @typedef Overlap
       * @type {object}
       * @property {('Overlap' | 'Coverage')} type
       * @property {OverlapRange[]} ranges
       */

      /**
       * @param {Range} range1
       * @param {Range} range2
       * @returns {(Overlap | null)}
       */
      function testOverlap(range1, range2) {
        if (range1[1] <= range2[0] || range1[0] >= range2[1]) {
          // No overlap
          return null;
        } else if (range1[0] <= range2[0] && range1[1] >= range2[1]) {
          // Complete coverage of range2 by range1
          const overlapIndices = [range2[0], range2[1]];
          const overlapInR1 = [
            overlapIndices[0] - range1[0],
            overlapIndices[1] - range1[0]
          ];
          const overlapInR2 = [0, overlapIndices[1] - range2[0]];
          return {
            type: "Coverage",
            ranges: [{
                range: range1,
                overlap: overlapInR1
              },
              {
                range: range2,
                overlap: overlapInR2
              }
            ]
          };
        } else if (range1[0] >= range2[0] && range1[1] <= range2[1]) {
          // Complete coverage of range1 by range2
          const overlapIndices = [range1[0], range1[1]];
          const overlapInR1 = [0, overlapIndices[1] - range1[0]];
          const overlapInR2 = [
            overlapIndices[0] - range2[0],
            overlapIndices[1] - range2[0]
          ];
          return {
            type: "Coverage",
            ranges: [{
                range: range2,
                overlap: overlapInR2
              },
              {
                range: range1,
                overlap: overlapInR1
              }
            ]
          };
        } else if (range1[1] >= range2[0] && range1[0] < range2[0]) {
          // Overlap on the left of range2
          const overlapIndices = [range2[0], range1[1]];
          const overlapInR1 = [
            overlapIndices[0] - range1[0],
            overlapIndices[1] - range1[0]
          ];
          const overlapInR2 = [0, overlapIndices[1] - range2[0]];
          return {
            type: "Overlap",
            ranges: [{
                range: range1,
                overlap: overlapInR1
              },
              {
                range: range2,
                overlap: overlapInR2
              }
            ]
          };
        } else if (range1[0] <= range2[1] && range1[1] > range2[1]) {
          // Overlap on the right of range2
          const overlapIndices = [range1[0], range2[1]];
          const overlapInR1 = [
            overlapIndices[0] - range1[0],
            overlapIndices[1] - range1[0]
          ];
          const overlapInR2 = [overlapIndices[0] - range2[0], range2[1] - range2[0]];
          return {
            type: "Overlap",
            ranges: [{
                range: range1,
                overlap: overlapInR1
              },
              {
                range: range2,
                overlap: overlapInR2
              }
            ]
          };
        }
      }

      // build array of text nodes and body string indicies
      var n;
      var walk = document.createTreeWalker(
        document.querySelector('#body'),
        NodeFilter.SHOW_TEXT
      );

      /**
       * @typedef NodeInfo
       * @type {object}
       * @property {Text} node TextNode from DOM. Note that html escape codes are decoded (&amp; -> &)
       * @property {string} text
       * @property {Range} range
       */

      /** @type {NodeInfo[]} */
      var textNodes = [];
      var traversedLength = 0;
      while (n = walk.nextNode()) {
        var encodedText = htmlEncode(n.wholeText);
        textNodes.push({
          node: n,
          text: encodedText,
          range: [traversedLength, traversedLength + encodedText.length]
        })
        traversedLength = traversedLength + encodedText.length
      }

      function markNode(node, text, start, stop) {

        var container = document.createElement('span')
        var beforeMark = htmlDecode(text.slice(0, start));
        var textToMark = htmlDecode(text.slice(start, stop));
        var afterMark = htmlDecode(text.slice(stop));

        var mark = document.createElement('mark');
        var textNodeToMark = document.createTextNode(textToMark);
        mark.appendChild(textNodeToMark);

        textNodeBeforeMark = document.createTextNode(beforeMark);
        textNodeAfterMark = document.createTextNode(afterMark);

        container.appendChild(textNodeBeforeMark);
        container.appendChild(mark);
        container.appendChild(textNodeAfterMark);

        node.replaceWith(container);
      }

      /**
       * @param {Range[]} ranges Ranges of text on the document to mark
       */
      function markTextRanges(ranges) {
        for (var i = 0; i < ranges.length; i++) {
          var range = ranges[i]

          for (var j = 0; j < textNodes.length; j++) {
            var textInfo = textNodes[j]
            var overlap = testOverlap(range, textInfo.range)
            if (overlap === null) continue
            if (overlap.ranges[0].range === textInfo.range) {
              var overlapRange = overlap.ranges[0].overlap
              markNode(textInfo.node, textInfo.text, overlapRange[0], overlapRange[1])
            } else {
              var overlapRange = overlap.ranges[1].overlap
              markNode(textInfo.node, textInfo.text, overlapRange[0], overlapRange[1])
            }
          }
        }
      }

      return {
        markTextRanges
      }
    })()
  </script>
</body>

</html>"
`;
